tools/unicode_properties_parse/format_width_estimate_intervals.cpp

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// The following code generates data for `_Width_estimate_intervals` in <format>.
+
+#include <charconv>
+#include <format>
+#include <fstream>
+#include <iostream>
+#include <regex>
+#include <vector>
+
+void verify_impl(bool test, int line, const char* msg) {
+    if (!test) {
+        std::cerr << "Error at line " << line << ": " << msg << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+#define verify(expr, msg) verify_impl((expr), __LINE__, (msg))
+constexpr const char* impl_assertion_failed = "impl assertion failed";
+
+struct range_u {
+    uint32_t from, to;
+    range_u(uint32_t f, uint32_t t) : from(f), to(t) {}
+    range_u(uint32_t v) : from(v), to(v) {}
+};
+
+enum class width_u : bool { is_1 = false, is_2 = true };
+
+class table_u {
+    // A valid Unicode code point won't exceed `max_u`.
+    static constexpr uint32_t max_u = 0x10'ffff;
+    std::vector<width_u> table;
+
+public:
+    table_u() : table(max_u + 1, width_u::is_1) {}
+    void fill_range(const range_u rng, const width_u width) {
+        const auto [from, to] = rng;
+        verify(from <= to && to <= max_u, impl_assertion_failed);
+        for (uint32_t u = from; u <= to; u++) {
+            table[u] = width;
+        }
+    }
+
+    void print_intervals() const {
+        // Print table for `_Width_estimate_intervals`.
+        using namespace std;
+        int c        = 0;
+        width_u last = table[0];
+        for (uint32_t u = 0; u <= max_u; u++) {
+            if (table[u] != last) {
+                cout << "0x" << hex << uppercase << u << "u, ";
+                if (++c == 12) {
+                    c = 0;
+                    cout << endl;
+                }
+            }
+            last = table[u];
+        }
+        cout << endl;
+    }
+
+    void print_clusters_1_vs_2(const table_u& other) const {
+        using namespace std;
+        vector<bool> cluster_table(max_u + 1, false);
+        for (uint32_t u = 0; u <= max_u; u++) {
+            if (table[u] == width_u::is_1 && other.table[u] == width_u::is_2) {
+                cluster_table[u] = true;
+            }
+        }
+
+        for (uint32_t u = 0; u <= max_u; u++) {
+            if (cluster_table[u]) {
+                uint32_t from = u;
+                uint32_t to   = from;
+                while (to + 1 <= max_u && cluster_table[to + 1]) {
+                    ++to;
+                }
+                if (from == to) {
+                    cout << hex << uppercase << "U+" << from << endl;
+                } else {
+                    cout << hex << uppercase << "U+" << from << "..U+" << to << endl;
+                }
+                u = to;
+            }
+        }
+    }
+};
+
+table_u get_table_cpp20() {
+    const range_u std_wide_ranges_cpp20[]{

```

Comment:
This is just a small generator so there's no reason to gold-plate it, but it seems like a useful time to point out that there's a major difference in performance between local arrays that are static vs non-static. Because distinct objects in C++ need distinct addresses in general, non-static local arrays will generate relatively horrible code where the compiler individually pushes each value onto the static. (The compiler can theoretically notice that nobody cares about the address and coalesce objects under the As If Rule, except that arrays are virtually always used in an address-taken way which prevents the compiler from noticing that nobody actually cares about the location.) In contrast, `static` arrays can generate nice read-only data. (We should make the constructors `constexpr` to enable this. Note that there is no inherent difference between `static const` and `static constexpr` arrays when the construction is constexpr-eligible, but it's just a nice consistent practice to use `static constexpr` at function scope for arrays.)

For individual variables, plain `constexpr` is fine and `static constexpr` generates no additional value, since we don't typically take their address.